source/distributed/tria.cc

```diff
@@ -1124,10 +1124,545 @@ namespace
+      // function and proceed its position to the next cell
+      // after each iteration.
+      auto it_qcrel     = quad_cell_relations.begin();
+      auto cit_dest_fix = dest_data_fix.cbegin() + offset;
+      for (; it_qcrel != quad_cell_relations.end();
+           ++it_qcrel, cit_dest_fix += cumulative_sizes_fix.back())
+        {
+          const auto &cell_status = std::get<1>(*it_qcrel);
+          const auto &dealii_cell = std::get<2>(*it_qcrel);
+
+          switch (cell_status)
+            {
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_PERSIST:
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_COARSEN:
+                unpack_callback(dealii_cell,
+                                cell_status,
+                                boost::make_iterator_range(cit_dest_fix,
+                                                           cit_dest_fix +
+                                                             size));
+                break;
+
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_REFINE:
+                unpack_callback(dealii_cell->parent(),
+                                cell_status,
+                                boost::make_iterator_range(cit_dest_fix,
+                                                           cit_dest_fix +
+                                                             size));
+                break;
+
+              case parallel::distributed::Triangulation<dim,
+                                                        spacedim>::CELL_INVALID:
+                // Skip this cell.
+                break;
+
+              default:
+                Assert(false, ExcInternalError());
+                break;
+            }
+        }
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    Triangulation<dim, spacedim>::DataTransfer::save(
+      const typename dealii::internal::p4est::types<dim>::forest
+        *         parallel_forest,
+      const char *filename) const
+    {
+      Assert(cumulative_sizes_fix.size() > 0,
+             ExcMessage("No data has been packed!"));
+
+      const std::string fname_fixed = std::string(filename) + "_fixed.data";
+
+      // ----- copied -----
+      // from DataOutInterface::write_vtu_parallel
+      // TODO: write general MPIIO interface
+      int myrank, nproc;
+      int ierr = MPI_Comm_rank(mpi_communicator, &myrank);
+      AssertThrowMPI(ierr);
+      ierr = MPI_Comm_size(mpi_communicator, &nproc);
+      AssertThrowMPI(ierr);
+
+      MPI_Info info;
+      ierr = MPI_Info_create(&info);
+      AssertThrowMPI(ierr);
+      MPI_File fh;
+      ierr = MPI_File_open(mpi_communicator,
+                           const_cast<char *>(fname_fixed.c_str()),
```

Comment:
Do you get an error without the `const_cast`? The man pages of `MPI_File_open` suggest that the argument passed here is a `const char *`: https://www.open-mpi.org/doc/v2.0/man3/MPI_File_open.3.php

Reply:
I do not get an error.
Will postpone it to the later pull request as well!